src/main/java/org/openstreetmap/atlas/generator/AtlasGenerator.java

```diff
@@ -284,243 +360,428 @@ public void start(final CommandMap command)
                     {
-                        // Generate the Atlas for this shard
-                        atlas = loader.load(countryName, shard);
+                        // Generate the raw Atlas for this shard
+                        atlas = loader.generateRawAtlas(countryName, shard);
                     }
                     catch (final Throwable e)
                     {
-                        throw new CoreException("Building Atlas {} failed!", name, e);
+                        throw new CoreException("Building raw Atlas {} failed!", name, e);
                     }
 
                     // Report on memory usage
-                    logger.info("Printing memory after loading Atlas {}", name);
+                    logger.info("Printing memory after loading raw Atlas {}", name);
                     Memory.printCurrentMemory();
+
                     // Output the Name/Atlas couple
                     final Tuple2<String, Atlas> result = new Tuple2<>(name, atlas);
                     return result;
-                });
+                }).filter(tuple -> tuple._2() != null);
 
-        // Filter out null Atlas.
-        final JavaPairRDD<String, Atlas> countryNonNullAtlasShardsRDD = countryAtlasShardsRDD
-                .filter(tuple -> tuple._2() != null);
+        // Persist the RDD to avoid re-computing and save each raw atlas
+        countryRawAtlasShardsRDD.cache().saveAsHadoopFile(
+                getAlternateSubFolderOutput(outputDirectory, RAW_ATLAS_FOLDER), Text.class,
+                Atlas.class, MultipleAtlasOutputFormat.class, new JobConf(configuration()));
+        logger.info("\n\n********** SAVED THE RAW ATLAS **********\n");
 
-        // Cache the Atlas
-        countryNonNullAtlasShardsRDD.cache();
-        logger.info("\n\n********** CACHED THE ATLAS **********\n");
+        return countryRawAtlasShardsRDD;
+    }
 
-        // Run the metrics
-        final JavaPairRDD<String, AtlasStatistics> statisticsRDD = countryNonNullAtlasShardsRDD
+    private JavaPairRDD<String, Atlas> sectionAndSaveRawAtlas(
+            final JavaPairRDD<String, Atlas> countrySlicedRawAtlasShardsRDD,
+            final CountryBoundaryMap boundaries, final Map<String, String> sparkContext,
+            final Map<String, String> loadingOptions, final String outputDirectory)
+    {
+        final JavaPairRDD<String, Atlas> countryAtlasShardsRDD = countrySlicedRawAtlasShardsRDD
                 .mapToPair(tuple ->
                 {
-                    final Counter counter = new Counter().withSharding(sharding);
-                    counter.setCountsDefinition(Counter.POI_COUNTS_DEFINITION.getDefault());
-                    final AtlasStatistics statistics;
+                    final Atlas atlas;
                     try
                     {
-                        statistics = counter.processAtlas(tuple._2());
+                        final AtlasLoadingOption atlasLoadingOption = buildAtlasLoadingOption(
```

Comment:
I am not sure I remember how the legacy code passes the `AtlasLoadingOption` around, but I believe it did not have to do this. We can review this at some point, there might be another way :)

Reply:
I don't think it passes it around, each slave builds it's own AtlasLoadingOption! Building it once and passing it around didn't work, since it will try to serialize the entire class and its contents - some of which aren't Serializable.

Reply:
I'll try to think of something a little more elegant, if you have ideas, please let me know!